@nrwl/next can't use 14.5.5 with ReactDOM 17.x

<!-- Please do your best to fill out all of the sections below! -->

## Current Behavior

At the moment if users running ReactDOM 17.x in their Next.js projects are going to face issues with building an application as it defaults to using ReactDOM 18.x, possibly due to changes introduced in https://github.com/nrwl/nx/commit/50afd2b6648ce967ed76437598315df016278a35. Setting `__NEXT_REACT_ROOT` to `false` doesn't solve the issue as it appears that Next.js only checks for [presence of any value](https://github.com/vercel/next.js/blob/canary/packages/next/client/index.tsx#L38) in `__NEXT_REACT_ROOT`.

## Expected Behavior

Setting `__NEXT_REACT_ROOT` based on installed React/ReactDOM versions would be ideal scenario.

## Steps to Reproduce

1. Bootstrap fresh project with `npx create-nx-workspace@latest` selecting `next.js` as application framework.
2. Revert React packages to use v17
3. Run a production build

Github repo - https://github.com/G07cha/react-17-bug

### Failure Logs

```
$ npm run build

> next-17-react-bug@0.0.0 build
> nx build


> nx run next-17-react-bug:build:production

info  - Loaded env from apps/next-17-react-bug/.env
info  - Skipping linting
info  - Checking validity of types
info  - Creating an optimized production build
Failed to compile.

../../node_modules/next/dist/client/index.js
Module not found: Can't resolve 'react-dom/client' in '<path-to-project>/next-17-react-bug/node_modules/next/dist/client'

Import trace for requested module:
../../node_modules/next/dist/client/next.js


 >  NX   Build failed because of webpack errors

   Pass --verbose to see the stacktrace.


 ——————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Ran target build for project next-17-react-bug (2s)

    ✖    1/1 failed
    ✔    0/1 succeeded [0 read from cache]
```

### Environment

```
   Node : 17.7.1
   OS   : darwin arm64
   npm  : 8.5.2

   nx : 14.5.6
   @nrwl/angular : Not Found
   @nrwl/cypress : 14.5.6
   @nrwl/detox : Not Found
   @nrwl/devkit : 14.5.6
   @nrwl/eslint-plugin-nx : 14.5.6
   @nrwl/express : Not Found
   @nrwl/jest : 14.5.6
   @nrwl/js : 14.5.6
   @nrwl/linter : 14.5.6
   @nrwl/nest : Not Found
   @nrwl/next : 14.5.6
   @nrwl/node : Not Found
   @nrwl/nx-cloud : Not Found
   @nrwl/nx-plugin : Not Found
   @nrwl/react : 14.5.6
   @nrwl/react-native : Not Found
   @nrwl/schematics : Not Found
   @nrwl/storybook : 14.5.6
   @nrwl/web : 14.5.6
   @nrwl/workspace : 14.5.6
   typescript : 4.7.4
   ---------------------------------------
   Local workspace plugins:
   ---------------------------------------
   Community plugins:
```
<!-- It's important for us to know the context in which you experience this behavior! -->
<!-- Please paste the result of `nx report` below! -->

--- a/packages/next/package.json
+++ b/packages/next/package.json
@@ -47,6 +47,7 @@
     "eslint-config-next": "^12.1.0",
     "fs-extra": "^10.1.0",
     "ignore": "^5.0.4",
+    "semver": "7.3.4",
     "ts-node": "10.9.1",
     "tsconfig-paths": "^3.9.0",
     "url-loader": "^4.1.1",

--- a/packages/next/src/executors/build/build.impl.ts
+++ b/packages/next/src/executors/build/build.impl.ts
@@ -1,14 +1,15 @@
 import 'dotenv/config';
-import { ExecutorContext } from '@nrwl/devkit';
+import { ExecutorContext, readJsonFile, workspaceLayout } from '@nrwl/devkit';
 import build from 'next/dist/build';
 import { join, resolve } from 'path';
 import { copySync, mkdir } from 'fs-extra';
+import { gte } from 'semver';
 import { directoryExists } from '@nrwl/workspace/src/utilities/fileutils';
 import {
   calculateProjectDependencies,
   DependentBuildableProjectNode,
 } from '@nrwl/workspace/src/utilities/buildable-libs-utils';
-import { workspaceLayout } from '@nrwl/devkit';
+import { checkAndCleanWithSemver } from '@nrwl/workspace/src/utilities/version-utils';
 
 import { prepareConfig } from '../../utils/config';
 import { createPackageJson } from './lib/create-package-json';
@@ -23,7 +24,14 @@ export default async function buildExecutor(
 ) {
   // Cast to any to overwrite NODE_ENV
   (process.env as any).NODE_ENV ||= 'production';
-  (process.env as any).__NEXT_REACT_ROOT ||= 'true';
+
+  // Set `__NEXT_REACT_ROOT` based on installed ReactDOM version
+  const packageJson = readJsonFile(join(context.root, 'package.json'));
+  const reactDomVersion = packageJson.dependencies['react-dom'];
+  const hasReact18 =
+    reactDomVersion &&
+    gte(checkAndCleanWithSemver('react-dom', reactDomVersion), '18.0.0');
+  (process.env as any).__NEXT_REACT_ROOT ||= hasReact18 ? 'true' : undefined;
 
   let dependencies: DependentBuildableProjectNode[] = [];
   const root = resolve(context.root, options.root);

